core/distributed/matrix.cpp

```diff
@@ -408,20 +406,24 @@ void Matrix<ValueType, LocalIndexType, GlobalIndexType>::apply_impl(
                            row_gatherer_->get_collective_communicator()
                                ->get_recv_size()),
                        dense_b->get_size()[1]};
-            recv_buffer_.init(exec, recv_dim);
-            host_recv_buffer_.init(exec->get_master(), recv_dim);
+            auto recv_buffer =
+                recv_buffer_.template get<b_value_type>(exec, recv_dim);
```

Comment:
shouldn't this be
```suggestion
                recv_buffer_.template get<x_value_type>(exec, recv_dim);
```
I think I would rather have the communication use different value types than the second advanced apply.

Reply:
I think the distributed RowGatherer does not support mixed precision.
It takes the value type from b and create a temporary clone on x if the type is not matched.
Also, the matrix can have the mixed precision.

Reply:
Yes, I looked a bit into it, and this sort of mixed precision is not allowed by the MPI standard. So my original comment was pointless.